src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1015,3 +1065,923 @@ def upgrade_agents(cmd, client, resource_group_name, cluster_name, kube_config=N
+    output_helm_values, error_helm_get_values = response_helm_values_get.communicate()
+    if response_helm_values_get.returncode != 0:
+        if ('forbidden' in error_helm_get_values.decode("ascii") or 'timed out waiting for the condition' in error_helm_get_values.decode("ascii")):
+            telemetry.set_user_fault()
+        telemetry.set_exception(exception=error_helm_get_values.decode("ascii"), fault_type=consts.Get_Helm_Values_Failed,
+                                summary='Error while doing helm get values azure-arc')
+        raise CLIError("Error while getting the helm values in the azure-arc namespace: " + error_helm_get_values.decode("ascii"))
+
+    output_helm_values = output_helm_values.decode("ascii")
+
+    try:
+        existing_values = yaml.safe_load(output_helm_values)
+        return existing_values
+    except Exception as e:
+        telemetry.set_exception(exception=e, fault_type=consts.Helm_Existing_User_Supplied_Value_Get_Fault,
+                                summary='Problem loading the helm existing values')
+        raise CLIError("Problem loading the helm existing values: " + str(e))
+
+
+def enable_features(cmd, client, resource_group_name, cluster_name, kube_config=None, kube_context=None,
+                    features=None, aad_client_id=None, aad_client_secret=None):
+    if features is None:
+        raise CLIError(consts.No_Features_Param_Provided.format("enable-features", "enable-features"))
+
+    logger.warning("Ensure that you have the latest helm version installed before proceeding.")
+    logger.warning("This operation might take a while...\n")
+
+    features = [x.lower() for x in features]
+    enable_cluster_connect, enable_azure_rbac, enable_cl = utils.check_features_to_update(features)
+
+    if enable_azure_rbac:
+        if (aad_client_id is None) or (aad_client_secret is None):
+            telemetry.set_user_fault()
+            telemetry.set_exception(exception='Client ID or secret not provided for AAD RBAC', fault_type=consts.Client_Details_Not_Provided_For_AAD_RBAC_Fault,
+                                    summary='Both aad authorization client id and client secret is required to enable AAD RBAC feature')
+            raise CLIError("Please provide both aad authorization client id and client secret to enable AAD RBAC feature")
+
+    if enable_cl:
+        enable_cl, custom_locations_oid = check_cl_registration_and_get_oid(cmd)
+        if not enable_cluster_connect and enable_cl:
+            enable_cluster_connect = True
+            features.append("cluster-connect")
+            logger.warning("Enabling 'custom-locations' feature will enable 'cluster-connect' feature too.")
+        if not enable_cl:
+            features.remove("custom-locations")
+            if len(features) == 0:
+                raise CLIError("Based on the inputs given and the warnings displayed above, there are no features to be enabled.")
```

Comment:
What is this for?

Reply:
When the user wanted only custom-locations to be enabled, and CL provider was not registered, so then no feature to update.

Reply:
"Based on the inputs given and the warnings displayed above, there are no features to be ena...." what is this message for?

---
In reply to: [596651884](https://github.com/Azure/azure-cli-extensions/pull/3001#discussion_r596651884) [](ancestors = 596651884)

Reply:
User wanted only CL to be enabled, validation failed for allowing that(warning given), So, nothing to be enabled.

Reply:
We should re word this part, no more features to enable.

---
In reply to: [596641779](https://github.com/Azure/azure-cli-extensions/pull/3001#discussion_r596641779) [](ancestors = 596641779)

Reply:
Just change it to - "Failed to enabled custom locations feature."

---
In reply to: [597661072](https://github.com/Azure/azure-cli-extensions/pull/3001#discussion_r597661072) [](ancestors = 597661072,596641779)